### PR TITLE
Unix sockets

### PR DESCRIPTION
**Pull request by ehmry**
Platform specific stuff is best avoided but Unix sockets should be used when they are available.

I've only tested on Linux.

### CODE REVIEW

File: src/sys/private/sockets_posix.nim [hunk 1]
```diff
@@ -164,14 +164,7 @@ proc handleAsyncConnectResult(fd: SocketFD) {.raises: [OSError].} =
...
 
-template tcpConnect() {.dirty.} =
-  let addressFamily =
-    when endpoint is IP4Endpoint:
-      AF_INET
-    elif endpoint is IP6Endpoint:
-      AF_INET6
-  let sock = makeSocket(addressFamily, SOCK_STREAM, IPPROTO_TCP)
-
+template socketConnect() {.dirty.} =
```
Comment (alaviss) [inline src/sys/private/sockets_posix.nim:167]: Can you make this take parameters normally? Aside from the implementations I would prefer if none of the templates are `{.dirty.}`

File: src/sys/private/sockets_posix.nim [hunk 2]
```diff
@@ -189,17 +182,19 @@ template tcpConnect() {.dirty.} =
...
     elif endpoint is IP6Endpoint:
       AF_INET6
-  var sock = makeSocket(addressFamily, SOCK_STREAM, IPPROTO_TCP, {sfNonBlock})
+  let sock = makeSocket(addressFamily, SOCK_STREAM, IPPROTO_TCP)
+  socketConnect()
+
+  # Take ownership of the socket from the handle
+  result = Conn[TCP] newSocket(sock)
 
+template socketAsyncConnect() {.dirty.} =
```
Comment (alaviss) [inline src/sys/private/sockets_posix.nim:197]: It's the same thing for this.

File: src/sys/sockets.nim [hunk 3]
```diff
@@ -883,3 +884,49 @@ proc accept*(l: AsyncListener[TCP]): tuple[conn: AsyncConn[TCP], remote: IPEndpo
 proc localEndpoint*(l: AsyncListener[TCP] | Listener[TCP]): IPEndpoint {.raises: [OSError].} =
   ## Obtain the local endpoint of `l`.
   tcpLocalEndpoint()
+
+proc connectUnix*(path: string): Conn[Unix]
+                {.raises: [OSError].} =
```
Comment (alaviss) [inline src/sys/sockets.nim:889]:
```suggestion
proc connectUnix*(path: string): Conn[Unix]
                 {.raises: [OSError].} =
```

As for style, please align the `{.` with `(` of the previous line.

File: src/sys/private/sockets_posix.nim [hunk 4]
```diff
@@ -409,3 +419,78 @@ template tcpLocalEndpoint() {.dirty.} =
...
+          when declared(accept4):
+            let flags = SOCK_CLOEXEC or SOCK_NONBLOCK
+            accept4(l.fd.SocketHandle, nil, nil, flags)
+          else:
+            accept(l.fd.SocketHandle, nil, nil)
+    if handle.fd == InvalidFD:
+      if errno == EAGAIN or errno == EWOULDBLOCK:
+        wait(l.fd, Event.Read)
+      else:
+       raise newOSError(errno, $Error.Accept)
```
Comment (alaviss) [inline src/sys/private/sockets_posix.nim:490]:
```suggestion
        raise newOSError(errno, $Error.Accept)
```

File: src/sys/private/sockets_posix.nim [hunk 5]
```diff
@@ -409,3 +419,78 @@ template tcpLocalEndpoint() {.dirty.} =
     result = IPEndpoint(kind: V6, v6: cast[IP6Endpoint](saddr))
   else:
     doAssert false, "Unexpected remote address family: " & $saddr.ss_family
+
+proc makeUnixSockaddr(path: string): Sockaddr_un =
+  result.sun_family = AF_UNIX.TSa_Family
+  if path.len >= Sockaddr_un_path_length:
+    raise newException(OSError, "socket path too long")
```
Comment (alaviss) [inline src/sys/private/sockets_posix.nim:426]: I think `ValueError` is more appropriate here. `OSError` implies the existence of an OS error code embedded, for which we have none.

File: src/sys/sockets.nim [hunk 6]
```diff
@@ -883,3 +884,49 @@ proc accept*(l: AsyncListener[TCP]): tuple[conn: AsyncConn[TCP], remote: IPEndpo
...
+  ## Create a connection to the Unix socket at `path`.
```
Comment (alaviss) [inline src/sys/sockets.nim:890]: Can you add a "Platform specific details" block about how this API does not support Windows at the moment and that the length limit varies between operating systems?

Then please replicate that block to all Unix sockets functions docs that take a path.

File: tests/sockets/tsockets.nim [hunk 7]
```diff
@@ -555,3 +562,66 @@ suite "TCP sockets":
...
+      let conn = connectUnix(path)
+
+      # Read from the server and verify the data
+      check conn.delimitedRead(Delimiter) == TestDelimitedData
+
+      # Collect the writer
+      joinThread thr
+
+      # Remove the socket
+      discard unlink(path)
```
Comment (alaviss) [inline tests/sockets/tsockets.nim:593]: This should be in a `defer` to handle error cases.

File: tests/sockets/tsockets.nim [hunk 8]
```diff
@@ -555,3 +562,66 @@ suite "TCP sockets":
...
+        # Retrieve the data
+        check conn.accumlatedRead(TestBufferedData.len + 1024) == TestBufferedData
+
+      # Run the test
+      runner()
+      # Run the IO queue to completion
+      run()
+
+      # Remove the socket
+      discard unlink(path)
```
Comment (alaviss) [inline tests/sockets/tsockets.nim:627]: Same here.

**Review by alaviss — COMMENTED** [review]
Thank you for contributing!

There are a couple small adjustments that I would like to see, but everything else looks great.

File: src/sys/private/sockets_posix.nim (hunk 5 above)
Comment (ehmry) [reply src/sys/private/sockets_posix.nim]: We can preemptively raise `ENAMETOOLONG` here.

File: src/sys/private/sockets_windows.nim [hunk 9]
```diff
@@ -599,3 +599,9 @@ proc initWinsock() =
 
 # Initialize winsock on program startup.
 initWinsock()
+
+template unixConnect() {.dirty.} =
+  {.error: "Unix sockets not available on this platform".}
```
Comment (alaviss) [inline src/sys/private/sockets_windows.nim:604]: Unfortunately I think this doesn't work...

File: src/sys/sockets.nim [hunk 10]
```diff
@@ -883,3 +884,82 @@ proc accept*(l: AsyncListener[TCP]): tuple[conn: AsyncConn[TCP], remote: IPEndpo
...
+  ##
+  ##
+  ## Unix sockets are currently only supported on POSIX-like
+  ## platforms (this does not include Windows).
+  ## Unix socket implementations can differ in the maximum
+  ## supported length of path elements as well as the total
+  ## path length.
+  ## On Linux the maximum path length is hardcoded to 107
+  ## characters.
+  unixConnect()
```
Comment (alaviss) [inline src/sys/sockets.nim:900]: While I'm not a big fan of doing this, I think the solution would be to put all Unix sockets functions in a `when defined(posix)` until we gain a Windows implementation

File: tests/sockets/tsockets.nim (hunk 8 above)
Comment (alaviss) [reply tests/sockets/tsockets.nim:626]: Please remember to put this into some `defer`

**Review by alaviss — APPROVED** [review]
Just a few changes to make this compile on Windows and we're gold

**Pull request merged**